torch/_functorch/aot_autograd.py

```diff
@@ -1648,6 +1648,52 @@ def same_dtype_views(view1, view2):
     return True
 
 
+
+def tensors_definitely_do_not_overlap(x, y):
+    if torch._C._are_definitely_non_overlapping(x, y):
+        # not overlapping (this covers the basic cases that eager mode uses)
+        return True
+
+    # expected inputs: two lists of intervals in increasing order. Don't need to have the same length
+    def intervals_intersect(xs, ys):
+        if len(xs) == 0 or len(ys) == 0:
+            return False
+        x_ = xs[0]
+        y_ = ys[0]
+        if max(0, min(x_[1], y_[1]) - max(x_[0], y_[0])) > 0:
+            return True
+        if x_[1] > y_[1]:
+            return intervals_intersect(xs, ys[1:])
+        else:
+            return intervals_intersect(xs[1:], ys)
+
+    # This cases is needed for the shampoo optimizer.
+    # We can compute a list of contiguous intervals and check if any of them intersect
+    if x.dim() == 2 and y.dim() == 2 and x.stride()[1] == 1 and y.stride()[1] == 1:
+        # Compute the (start, end) range of contiguous memory of each region of the tensor
+        x_ranges = [
+            (x.data_ptr() + i * x.stride()[1], x.data_ptr() + i * x.stride()[1] + x.shape[1])
+            for i in range(x.shape[0] - 1)
+        ]
+        y_ranges = [
+            (y.data_ptr() + i * y.stride()[1], y.data_ptr() + i * y.stride()[1] + y.shape[1])
+            for i in range(y.shape[0] - 1)
+        ]
+        if not intervals_intersect(x_ranges, y_ranges):
+            return True
+    return False
+
+def compute_overlapping_inputs(fwd_inputs, aliased_input_indices):
+    actual_aliased_indices = set()
+    for j in range(len(aliased_input_indices)):
+        for i in range(j):
```

Comment:
So the idea here is that this "check if there is any real overlapping memory" happens in two stages:

Stage 1: We take all 200 of the graph inputs, and bucket them based on their `.untyped_storage()`. For each bucket, we keep track of a set of "what are (the indices of) the graph inputs that all share a storage. Say that out of our 200 graph inputs, one bucket is [37, 38, 39]

Stage 2: We take each of those buckets of indices, and figure out if any of them have "true overlapping memory". So `aliased_input_indices` here would refer to `[37, 38, 39]`, and we check pair-wise if any pair of tensors is overlapping.

Two reasons that I set it up this way:

(1) Stage 1 already exists, so it was easier to add a separate independent stage in for the extra overlapping memory checks

(2) In theory, doing this "check pair-wise if any pair of tensors has overlapping memory" can be pretty expensive, since (a) I saw that shampoo had 10-20 params that all shared a storage (and we have to do O(n^2) overlapping memory checks), and (b) the checks themselves can be slow (like this overlapping-interval logic that I added).

I agree the two stage logic is a bit complicated, but lmk if this sounds reasonable to you!

Reply:
That sounds good, Thanks for the details!